### ISSUE TITLE
Add regexp function - regexp_matches()

### ISSUE THREAD

**Issue by jonathanc-n** [apache/datafusion#13052]
### Describe the solution you'd like

regexp_match() is already implemented, so we just need to look into building on that. 

### Additional context

Discussed in #11946

**Comment by jonathanc-n** [comment]
take

**Comment by jonathanc-n** [comment]
@jcsherin Do you think it'd be preferred to have regexp_matches on the same page as regexp_match or should i create a new file for it?

**Comment by jcsherin** [comment]
Yes, it should be possible to share code between `regexp_match` and `regexp_matches` as they are similar in functionality.

**Comment by jonathanc-n** [comment]
I was thinking that it may be better for regexp_matches to be in a separate file as it would probably just be easier to maintain + matches introduces some additional complications to implementation. It would've probably been easier to have created the regexp_matches then implement match afterwards on top of it.

**Comment by jonathanc-n** [comment]
@jcsherin Sorry, a little new to the arrow/datafusion ecosystem. If i want to add a regexp_matches implementation, would i have to push that to arrow, and wait for the next version to come out?

**Comment by jcsherin** [comment]
I think it is easier (less friction) to first, get something working here (with docs, test coverage) and iterate quickly over feedback from other contributors. Then it will become evident which parts are better moved upstream to arrow. So you can get this working completely here.

If you run into any roadblocks, more experienced (not me) contributors will help you out 🤝
